### PR TITLE
ENH: geography constructor from geoarrow

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
~This depends on https://github.com/paleolimbot/s2geography/pull/23~

This also introduces the question: how to we deal with functions that depend on a specific version of s2geography? Short term we can just ensure we do a upstream release and require the latest version, but longer term we might need to have some mechanism for this? -> added some version definitions that can be used in C++

Some other notes:

- This PR adds a `from_geoarrow` that accepts any Arrow-compatible array object (through the [Arrow PyCapsule interface](https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html)). As a result, this is _not_ a numpy vectorized function that will take any dimension of input, it's purely a 1D -> 1D function.
  - At the moment this PR also only adds support for `__arrow_c_array__`, but probably should also support `__arrow_c_stream__`? (given that a lot of other libraries will often only implement that)
- The current upstream implementation of the `s2geography::geoarrow::Reader` already supports WKT, WKB and native (nested coordinates) encoding, so those work out of the box here as well.  
  Right now I only accept Arrow input with an extension type and let the Reader figure out the encoding, but we could also add a argument to let the user specify "WKT"/"WKB" and then accept a plain Arrow object (without geoarrow extension type)
- I did try to add a specific `from_wkt` that _is_ a normal vectorized function, but the problem is that pybind11's `vectorize` doesn't seem to like `std::string` argument input for a vectorized arg. -> this is done in https://github.com/benbovy/spherely/pull/50

### CODE REVIEW

File: src/geoarrow.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            to be linear on the plane. In that case, additional points will
+            be added to the line while creating the geography objects, to
+            ensure every point is within 100m of the original line.
+            By default (False), it is assumed that the edges are spherical
+            (i.e. represent the shortest path on the sphere between two points).
+        tessellate_tolerance : float, default 100.0
+            The maximum distance in meters that a point must be moved to
+            satisfy the planar edge constraint. This is only used if `planar`
+            is set to True.
+        geometry_encoding : str, default None
```
Comment (benbovy) [inline src/geoarrow.cpp:109]:
```suggestion
        geography_encoding : str, default None
```

Not sure which is best, though.

**Review by benbovy — APPROVED** [review]
Just left a couple of minor comments.

I'm not very familiar with (geo)arrow but this seems good to go to me.

File: tests/test_geoarrow.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    arr = ga.as_geoarrow([polygon_with_bad_hole_wkt])
+
+    result = spherely.from_geoarrow(arr)
+    assert (
+        str(result[0])
+        == "POLYGON ((20 35, 10 30, 10 10, 30 5, 45 20, 20 35), (20 15, 20 25, 30 20, 20 15))"
+    )
+
+    # if we force to not orient, we get an error
+    with pytest.raises(RuntimeError, match="Inconsistent loop orientations detected"):
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply tests/test_geoarrow.py:79]: Just catch and reraise with the same error?

File: src/geoarrow.cpp (hunk 1 above)
Comment (jorisvandenbossche) [reply src/geoarrow.cpp:109]: Yeah, given this is for geoarrow IO and in geoarrow we generally use the "geometry" term, I thought to keep that as well for the keyword ... But either way it is not ideal.

File: tests/test_geoarrow.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (benbovy) [reply tests/test_geoarrow.py:79]:
```cpp
try {
       // call s2geography API
} catch( const std::exception &ex ) {
       throw py::value_error(ex.what());
}
``` 

I haven't looked where this should be inserted, though.

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply tests/test_geoarrow.py]: That seems to work!

**Comment by benbovy** [comment]
Thanks @jorisvandenbossche

**Pull request merged**